src/EFCore.PG/Query/ExpressionTranslators/Internal/NpgsqlDateAddTranslator.cs

```diff
@@ -86,32 +87,48 @@ public virtual Expression Translate(MethodCallExpression methodCallExpression)
             if (!MethodInfoDatePartMapping.TryGetValue(methodCallExpression.Method, out var datePart))
                 return null;
 
-            // Ideally we would use Expression.Add() to have a simple plus-sign expression generated for us.
-            // But to represent an interval we'd need to use a TimeSpan, which does not represent months/years.
-            // We could use the provider-specific NpgsqlTimeSpan but it's not currently supported and is somewhat
-            // ugly, so we just generate the expression ourselves with CustomBinaryExpression
+            // When we have a constant, we create the interval via INTERVAL '1 days'. But this expression cannot be
+            // parameterized, so we use CAST($1 || ' days' AS interval) instead
+            // Note that a make_interval() function also exists, but accepts only int (for all fields except for
+            // seconds), so we don't use it.
+
+            Expression interval = null;
 
-            Expression instance = methodCallExpression.Object;
-            Expression amountToAdd = methodCallExpression.Arguments[0];
+            var instance = methodCallExpression.Object;
+            var amountToAdd = methodCallExpression.Arguments[0];
 
             if (instance is null || amountToAdd is null)
                 return null;
 
-            if (amountToAdd is ConstantExpression amount &&
-                amount.Type == typeof(double) &&
-                ((double)amount.Value >= int.MaxValue ||
-                 (double)amount.Value <= int.MinValue))
-                return null;
+            if (amountToAdd is ConstantExpression constantExpression)
+            {
+                if (constantExpression.Type == typeof(double) &&
+                    ((double)constantExpression.Value >= int.MaxValue ||
+                     (double)constantExpression.Value <= int.MinValue))
+                {
+                    return null;
+                }
+
+                interval = new SqlFragmentExpression($"INTERVAL '{amountToAdd} {datePart}'");
+            }
+            else
+            {
+                interval = new ExplicitCastExpression(
+                    new CustomBinaryExpression(amountToAdd, Expression.Constant(' ' + datePart), "||", typeof(string)),
+                    typeof(TimeSpan)
```

Comment:
Will this `typeof(TimeSpan)` cause issues for the NodaTime plugin?

Reply:
This is the specifically translator for the built-in `DateTime.Add()` - it doesn't really need  to be NodaTime compatible. I don't think we need to aim for interoperability between the built-in date/time types and NodaTime, especially as such interoperability doesn't actually exist at the .NET level (e.g. you can't add a BCL `DateTime` and a NodaTime `Period`).

In NodaTime things are in theory much simpler - NodaTime simply defines arithmetic operators rather than having special `Add*` methods, so you can simply do `LocalDate` + `Period`. EF Core automatically translates all operators to SQL (the opposite of method calls, which need to be manually and explicitly translated), so it should all work out of the box without any work on our side. The `NodaTimeTest.Add_Days` is supposed to test exactly this, but is failing for some reason (need to check(.

Reply:
Ope. I mistook this for being in the SQL query translator... my mistake!